compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirUninitializedEnumChecker.kt

```diff
@@ -0,0 +1,124 @@
+                    reporter.reportOn(source, FirErrors.UNINITIALIZED_VARIABLE, referencedDeclaration.symbol, context)
+                } else {
+                    // enum class EnumCompanion2(...) {
+                    //   INSTANCE(<!>foo<!>())
+                    //   companion object {
+                    //     fun foo = ...
+                    //   }
+                    // }
+                    // <!>Companion<!>.foo() v.s. <!>foo<!>()
+                    if ((expression.explicitReceiver as? FirResolvedQualifier)?.symbol?.fir == enumClass.companionObject) {
+                        reporter.reportOn(
+                            expression.explicitReceiver!!.source,
+                            FirErrors.UNINITIALIZED_ENUM_COMPANION,
+                            enumClass.symbol,
+                            context
+                        )
+                    } else {
+                        reporter.reportOn(source, FirErrors.UNINITIALIZED_ENUM_COMPANION, enumClass.symbol, context)
+                    }
+                }
+            }
+
+        }
+
+        // The enum entries of an enum class
+        if (referencedDeclaration in enumEntries) {
+            val callee = referencedDeclaration as FirEnumEntry
+            // Uninitialized from the point of view of members of that enum class
+            if (accessedContext in enumMemberProperties) {
+                // From KT-6054
+                // enum class MyEnum {
+                //   A, B;
+                //   val x = when(this) {
+                //     <!>A<!> -> ...
+                //     <!>B<!> -> ...
+                //   }
+                // }
+                reporter.reportOn(source, FirErrors.UNINITIALIZED_ENUM_ENTRY, callee.symbol, context)
+            }
+            // enum class D(...) {
+            //   D1(<!>D2<!>),
+            //   D2(...)
+            // }
+            accessedContext.correspondingEnumEntry(enumEntries)?.let { caller ->
+                var precedingEntry : FirEnumEntry? = null
+                enumEntries.forEach {
+                    if (precedingEntry != null) return@forEach
+                    if (it == callee || it == caller) {
+                        precedingEntry = it
+                    }
+                }
+                if (precedingEntry == caller) {
+                    reporter.reportOn(source, FirErrors.UNINITIALIZED_ENUM_ENTRY, callee.symbol, context)
+                }
+            }
+        }
+    }
+
+    private fun FirDeclaration.correspondingEnumEntry(enumEntries: Collection<FirEnumEntry>): FirEnumEntry? {
+        if (!isEnumEntryInitializer) return null
+        // Note that we need to compare fully qualified names here. For example,
+        //   enum class EnumCompanion3(...) {
+        //     INSTANCE(EnumCompanion2.foo())
+        //   }
+        // If we only compare entry name, INSTANCE, we may incorrectly return EnumCompanion2.INSTANCE for EnumCompanion3.INSTANCE init.
```

Comment:
To avoid this altogether, can we check if the accessed context is within the same enum class as the callee?

Reply:
Yeah, that is added to handle a case like an anonymous function you came up with, and this part is no longer necessary.